src/azul/plugins/repository/tdr/__init__.py

```diff
@@ -197,10 +200,19 @@ def list_sources(self,
                      authentication: Optional[Authentication]
                      ) -> List[TDRSourceRef]:
         tdr = self._user_authenticated_tdr(authentication)
+        try:
+            snapshots = tdr.snapshot_names_by_id()
+        except UnauthorizedError:
+            if authentication is None:
+                raise
+            else:
+                # Fall back to anonymous access if the user-provided credentials
+                # lack authorization
+                return self.list_sources(None)
```

Comment:
I think we should be a little more specific here. An expired or syntactically invalid token should result in an exception, as before. There must be a way to check a token's validity against Google APIs. 

Also: I'm not sure this here is before or after the caching but we definitely don't want to create cache lines for invalid tokens.

Reply:
This is before the caching. The result of this method is used to populate the cache.